src/node/upload.ts

```diff
@@ -0,0 +1,150 @@
+} from '../common/upload.js';
+import {
+  TurboAuthenticatedUploadServiceConfiguration,
+  TurboUploadDataItemResponse,
+  TurboUploadFolderParams,
+  TurboUploadFolderResponse,
+  isNodeUploadFolderParams,
+} from '../types.js';
+
+export class TurboAuthenticatedNodeUploadService extends TurboAuthenticatedBaseUploadService {
+  constructor({
+    url = defaultUploadServiceURL,
+    retryConfig,
+    signer,
+    logger,
+    token,
+  }: TurboAuthenticatedUploadServiceConfiguration) {
+    super({ url, retryConfig, logger, token, signer });
+  }
+
+  private async getAbsoluteFilePathsFromFolder(
+    folderPath: string,
+  ): Promise<string[]> {
+    const absoluteFilePaths: string[] = [];
+
+    // Walk the directory and add all file paths to the array
+    const files = await promises.readdir(folderPath);
+    for (const file of files) {
+      const absoluteFilePath = join(folderPath, file);
+      const stat = await promises.stat(absoluteFilePath);
+      if (stat.isDirectory()) {
+        absoluteFilePaths.push(
+          ...(await this.getAbsoluteFilePathsFromFolder(absoluteFilePath)),
+        );
+      } else {
+        absoluteFilePaths.push(absoluteFilePath);
+      }
+    }
+    return absoluteFilePaths;
+  }
+
+  async uploadFolder(
+    params: TurboUploadFolderParams,
+  ): Promise<TurboUploadFolderResponse> {
+    if (!isNodeUploadFolderParams(params)) {
+      throw new Error('folderPath is required for node uploadFolder');
+    }
+    const {
+      folderPath,
+      indexFile,
+      dataItemOpts,
+      signal,
+      maxConcurrentUploads = 5,
+      throwOnFailure = true,
+    } = params;
+
+    const paths: Record<string, { id: string }> = {};
+    const fileResponses: TurboUploadDataItemResponse[] = [];
+    const errors: Error[] = [];
+
+    const absoluteFilePaths =
+      await this.getAbsoluteFilePathsFromFolder(folderPath);
+
+    const limit = pLimit(maxConcurrentUploads);
+
+    const uploadFile = async (absoluteFilePath: string) => {
+      const mimeType = lookup(absoluteFilePath);
+      const contentType =
+        mimeType === false ? 'application/octet-stream' : mimeType;
+
+      try {
+        const result = await this.uploadFile({
+          fileStreamFactory: () => createReadStream(absoluteFilePath),
+          fileSizeFactory: () => statSync(absoluteFilePath).size,
+          signal,
+          dataItemOpts: {
+            ...dataItemOpts,
+            tags: [
+              ...(dataItemOpts?.tags ?? []),
+              { name: 'Content-Type', value: contentType },
+            ],
+          },
+        });
+        fileResponses.push(result);
+        const relativePath = absoluteFilePath.replace(folderPath + '/', '');
+        paths[relativePath] = { id: result.id };
+      } catch (error) {
+        if (throwOnFailure) {
+          throw error;
+        }
+        this.logger.error(`Error uploading file: ${absoluteFilePath}`);
+        this.logger.error(error);
+        errors.push(error);
+      }
```

Comment:
i've started to become a fan of the await/catch pattern - particularly in these nested try blocks. what do you think?

```suggestion
        const result = await this.uploadFile({
          fileStreamFactory: () => createReadStream(absoluteFilePath),
          fileSizeFactory: () => statSync(absoluteFilePath).size,
          signal,
          dataItemOpts: {
            ...dataItemOpts,
            tags: [
              ...(dataItemOpts?.tags ?? []),
              { name: 'Content-Type', value: contentType },
            ],
          },
        }).catch((error: any) => {
             this.logger.error(`Error uploading file: ${absoluteFilePath}`, {
                  message: error?.message,
                  stack: error?.stack
             });
             if (throwOnFailure) {
                throw error;
             }
             errors.push(error);
        });
        fileResponses.push(result);
        const relativePath = absoluteFilePath.replace(folderPath + '/', '');
        paths[relativePath] = { id: result.id };
```

Reply:
that gives some issues on types 
<img width="1230" alt="image" src="https://github.com/user-attachments/assets/81b279f9-657b-466a-82d1-44bfeb99341c">

we need it to be a then chain on the other parts. and then we're mixing await and then a bunch, might as well make this a then.catch function

Reply:
I converted to then/catch, but the linter actually wanted me to convert it back to async function with try/catch. just reverted this for now. overall I think mixing `await` with `.catch()` makes flow control a bit tougher to reason about